sdk/metric/pipeline.go

```diff
@@ -134,3 +138,111 @@ func (p *pipeline) produce(ctx context.Context) (metricdata.ResourceMetrics, err
 		ScopeMetrics: sm,
 	}, nil
 }
+
+// pipelineRegistry manages creating pipelines, and aggregators.  The meters can
+// retrieve new aggregators from a registry.
+type pipelineRegistry[N int64 | float64] struct {
+	views     map[Reader][]view.View
+	pipelines map[Reader]*pipeline
+}
+
+func newPipelineRegistries(views map[Reader][]view.View) (*pipelineRegistry[int64], *pipelineRegistry[float64]) {
+	pipelines := map[Reader]*pipeline{}
+	for rdr := range views {
+		pipe := &pipeline{}
+		rdr.register(pipe)
+		pipelines[rdr] = pipe
+	}
+	return &pipelineRegistry[int64]{
+			views:     views,
+			pipelines: pipelines,
+		}, &pipelineRegistry[float64]{
+			views:     views,
+			pipelines: pipelines,
+		}
+}
+
+// createAggregators will create all backing aggregators for an instrument.
+// It will return an error if an instrument is registered more than once.
+// Note: There may be returned aggregators with an error.
+func (reg *pipelineRegistry[N]) createAggregators(inst view.Instrument, instUnit unit.Unit) ([]internal.Aggregator[N], error) {
+	var aggs []internal.Aggregator[N]
+
+	errs := &multierror{}
+	for rdr, views := range reg.views {
+		pipe := reg.pipelines[rdr]
+		rdrAggs := createAggregators[N](rdr, views, inst)
+		for inst, agg := range rdrAggs {
+			err := pipe.addAggregator(inst.Scope, inst.Name, inst.Description, instUnit, agg)
+			if err != nil {
+				errs.append(err)
+			}
+			aggs = append(aggs, agg)
+		}
+	}
+	return aggs, errs.errorOrNil()
+}
+
+func (reg *pipelineRegistry[N]) registerCallback(fn func(context.Context)) {
+	for _, pipe := range reg.pipelines {
+		pipe.addCallback(fn)
+	}
+}
+
+type multierror struct {
+	errors []string
+}
+
+func (m *multierror) errorOrNil() error {
+	if len(m.errors) == 0 {
+		return nil
+	}
+	return fmt.Errorf(strings.Join(m.errors, "; "))
+}
+func (m *multierror) append(err error) {
+	m.errors = append(m.errors, err.Error())
+}
+
+func createAggregators[N int64 | float64](rdr Reader, views []view.View, inst view.Instrument) map[view.Instrument]internal.Aggregator[N] {
+	aggs := map[view.Instrument]internal.Aggregator[N]{}
+	for _, v := range views {
+		inst, match := v.TransformInstrument(inst)
+		if inst.Aggregation == nil {
+			inst.Aggregation = rdr.aggregation(inst.Kind)
+		}
+		if match {
+			if _, ok := aggs[inst]; ok {
+				continue
+			}
+			agg := createAggregator[N](inst.Aggregation, rdr.temporality(inst.Kind))
```

Comment:
It sounds like other SIGs are adding guards here to prevent certain aggregators being created for certain instruments: https://cloud-native.slack.com/archives/C01NP3BV26R/p1658955998653879?thread_ts=1658938194.862009&cid=C01NP3BV26R

That thread also points out that this is under-specified, but I think we should follow suit until it is resolved. I wouldn't want the situation where we allowed behavior that the specification evolves to restrict and we have to remove it.

Reply:
If we are going to follow what java does, then from their code I made this mapping:

| Instrument Kind      | Drop | LastValue | Sum | Histogram |
|----------------------|------|-----------|-----|-----------|
| Sync Counter         | X    |           | X   | X         |
| Sync UpDown Counter  | X    |           | X   |           |
| Sync Histogram       | X    |           | X   | X         |
| Async Counter        | X    |           | X   |           |
| Async UpDown Counter | X    |           | X   |           |
| Async Gauge          | X    | X         |     |           |

Is that what you would expect, or should we do something different?  

Just note I don't think python has this kind of restriction.

Reply:
Looks reasonable to me. @jmacd thoughts?